src/expr.rs

```diff
@@ -221,38 +222,108 @@ impl<'a> FmtVisitor<'a> {
         format!("({})", item_str)
     }
 
+    fn rewrite_binary_op(&mut self,
+                         op: &ast::BinOp,
+                         lhs: &ast::Expr,
+                         rhs: &ast::Expr,
+                         width: usize,
+                         offset: usize)
+        -> String
+    {
+        let operator_str = self.snippet(op.span);
+
+        // 1 = space between lhs expr and operator
+        let mut result = self.rewrite_expr(lhs, width - 1 - operator_str.len(), offset);
+
+        result.push(' ');
+        result.push_str(&operator_str);
+
+        let remaining_width = match result.rfind('\n') {
+            Some(idx) => {
+                let space = config!(max_width) + idx;
+                let used = offset + result.len();
+                if space >= used {
+                    space - used
+                } else {
+                    0
+                }
+            },
+            None => {
+                if width <= result.len() {
+                    0
+                } else {
+                    width - result.len()
+                }
+            }
+        };
+
+        // Get "full width" rhs and see if it fits on the current line. This
+        // usually works fairly well since it tends to place operands of
+        // operations with high precendence close together.
+        let rhs_result = self.rewrite_expr(rhs, width, offset);
+
+        if rhs_result.len() > remaining_width {
+            result.push('\n');
+            result.push_str(&make_indent(offset));
+        } else {
+            result.push(' ');
+        };
+
+        result.push_str(&rhs_result);
+        result
+    }
+
+    fn rewrite_unary_op(&mut self, op: &ast::UnOp, expr: &ast::Expr, width: usize, offset: usize)
+        -> String
+    {
+        // For some reason, an UnOp is not spanned like BinOp!
+        let operator_str = match *op {
+            ast::UnOp::UnUniq => "&",
+            ast::UnOp::UnDeref => "*",
+            ast::UnOp::UnNot => "!",
+            ast::UnOp::UnNeg => "-"
+        };
+
+        operator_str.to_owned() + &self.rewrite_expr(expr, width - operator_str.len(), offset)
```

Comment:
I prefer using format! rather than concatenation - I think this particular case is OK, but in general you are more likely to get better allocation behaviour using format!